### PR TITLE
Faster lub

### PR DESCRIPTION
**Pull request by jvilk-stripe**
A continuation of https://github.com/sorbet/sorbet/pull/3241 which makes lub much faster for large `OrType`s.

### Motivation

Sorbet is slow to typecheck files that `case` over a large enum with hundreds of members. That slowness stems from lub.

This change results in a large speedup on these files (10X+).

### Test plan

See included automated tests.

### CODE REVIEW

File: core/types/subtyping.cc [hunk 1]
```diff
@@ -221,7 +258,9 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
         newTargs.reserve(indexes.size());
         // code below inverts permutation of type params
         int j = 0;
-        bool changed = false;
+        bool changedFromT2 = false;
+        // If klasses are equal, then it's possible that t1s <: t2s.
+        bool changedFromT1 = a1->klass != a2->klass;
```
Comment (jvilk-stripe) [inline core/types/subtyping.cc:263]: This is the correctness fix.

Before this change:
* `lub(T::Array[T.untyped], T::Array[X])` would return a `T::Array[T.untyped]` that is pointer equivalent with the first parameter.
* `lub(T::Array[X], T::Array[T.untyped])` would return a fresh `T::Array[T.untyped]` that is not pointer equivalent with the second parameter.

After this change:
* Both `lub(T::Array[T.untyped], T::Array[X])` and `lub(T::Array[X], T::Array[T.untyped])` return the same `T::Array[T.untyped]` pointer passed in to lub.

**Comment by jvilk-stripe** [comment]
This PR exposes new errors on Stripe's codebase; some look legitimate, others look false.

So I should probably complete that investigation prior to a full review.

**Comment by jvilk-stripe** [comment]
This PR is now ready for review.

File: test/testdata/infer/lub_and_glb_corner_cases.rb [hunk 2]
```diff
@@ -51,4 +51,27 @@ def generic_class_glb_collapse
...
+    if b == 1
+      xs = T::Array[T.nilable(String)].new
+    elsif b == 2
+      xs = nil
+    else
+      xs = T::Array[T.untyped].new
+    end
+    
+    # T.any(nil, T::Array[T.nilable(String)], T::Array[T.untyped]) => T.nlable(T::Array[T.untyped])
+    T.assert_type!(xs, T.nilable(T::Array[T.untyped]))
```
Comment (elliottt) [inline test/testdata/infer/lub_and_glb_corner_cases.rb:65]: Could you use `T.reveal_type` here instead?

File: test/testdata/infer/lub_and_glb_corner_cases.rb [hunk 3]
```diff
@@ -51,4 +51,27 @@ def generic_class_glb_collapse
...
+  end
+
+  def generic_class_nilable_array_tuple_lub
+    a = T.let(nil, T.nilable(T::Array[Float]))
+    b = T.let(nil, T.nilable(T::Array[Integer]))
+    T.assert_type!([a, b], [T.nilable(T::Array[Float]), T.nilable(T::Array[Integer])])
```
Comment (elliottt) [inline test/testdata/infer/lub_and_glb_corner_cases.rb:71]: Same for this one.

File: test/testdata/infer/fast_big_lubs.rb [hunk 4]
```diff
@@ -0,0 +1,533 @@
+# typed: true
```
Comment (elliottt) [inline test/testdata/infer/fast_big_lubs.rb:1]: Could you move this file into `test/testdata/perf`? We have a few other performance regression tests there already.

**Review by elliottt — APPROVED** [review]
The change to `Types::lub` makes sense to me, thanks for the additional test cases! Were there any other changes to @DarkDimius's original PR that weren't obvious? It looked like there were some small changes to `lubDistributeOr`, but it wasn't clear if the ended up affecting its behavior.

I'd also be interested in seeing a build with the new errors raised, when you get a chance 👍

**Comment by jvilk-stripe** [comment]
Links to internal Stripe builds:

* With typechecking errors: https://cibot.corp.stripe.com/builds/bui_Hx3kQn7Zor4oS3
* Clean build with fixes from a branch: https://cibot.corp.stripe.com/builds/bui_HzaU9panlyfxC9

File: test/testdata/infer/fast_big_lubs.rb (hunk 4 above)
Comment (jvilk-stripe) [reply test/testdata/infer/fast_big_lubs.rb:1]: Good idea; can do

**Comment by jvilk-stripe** [comment]
>Were there any other changes to @DarkDimius's original PR that weren't obvious? It looked like there were some small changes to `lubDistributeOr`

Nope. See: https://github.com/sorbet/sorbet/compare/98939cc..f9974ea

I added some comments, made an `if` into an `else if`, and fixed a bug in the metric Dmitry's original code was reporting.

File: test/testdata/infer/lub_and_glb_corner_cases.rb (hunk 2 above)
Comment (jvilk-stripe) [reply test/testdata/infer/lub_and_glb_corner_cases.rb:65]: Talked in DM; `reveal_type` shows the actual type Sorbet sees and doesn't reduce it, so it's better for testing lub. `T.assert_type!` calls into lub and may reduce the type further.

**Pull request merged**